File: Programmers/Lv_3/정수삼각형/정수삼각형_지민성.py
def solution(triangle):
    for row in range(len(triangle)):
        if row == 0:
            continue
        for i in range(row + 1):
            if i == 0:
                triangle[row][i] = triangle[row][i] + triangle[row - 1][i]
            elif i == row:
                triangle[row][i] = triangle[row][i] + triangle[row - 1][i - 1]
            else:
                triangle[row][i] = triangle[row][i] + max(triangle[row - 1][i - 1], triangle[row - 1][i])
    return max(triangle[-1])


print(solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]))

File: Programmers/Lv_3/정수삼각형/정수삼각형_지민성.java
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        for (int i = 1; i < triangle.length; i++) {
            int k = triangle[i].length;
            for (int j = 0; j < k; j++) {
                if (j == 0) {
                    triangle[i][j] = triangle[i][j] + triangle[i - 1][j];
                } else if (j == k - 1) {
                    triangle[i][j] = triangle[i][j] + triangle[i - 1][j - 1];
                } else {
                    triangle[i][j] = triangle[i][j] + Math.max(triangle[i - 1][j - 1], triangle[i - 1][j]);
                }
            }
        }
        for (int[] line : triangle)
            for (int i : line)
                answer = Math.max(answer, i);
        return answer;
    }
}